### PR TITLE
Dask tutorial with operator created only once

### PR DESCRIPTION
**Pull request by ofmla**
Some modifications were introduced to tutorial 04. Now the Operator is created only once and it is reused for the modeling of all the shots and the gradient computation. In this way, there should not be memory issues.

### CODE REVIEW

File: examples/seismic/tutorials/04_dask.ipynb [hunk 1]
```diff
@@ -178,41 +180,40 @@
...
-    "    rec_coordinates[:, 0] = 980. # 20m from the right end\n",
+    "    geometry.src_positions[0,:]=[30, shot_id*1000./(param['nshots']-1)]\n",
+    "    true_d, _, _ = op.forward()\n",
+    "    dump_shot_data(shot_id, true_d, geometry)\n",
     "\n",
+    "def generate_shotdata(src_coor,rec_coor):\n",
+    "    true_model = get_true_model()\n",
     "    # Geometry \n",
-    "    geometry = AcquisitionGeometry(true_model, rec_coordinates, src_coordinates,\n",
+    "    geometry = AcquisitionGeometry(true_model, rec_coor, src_coor,\n",
```
Comment (FabioLuporini) [inline examples/seismic/tutorials/04_dask.ipynb]: for homogeneity with the other notebooks, I'd keep the original variable names

Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: OK, Do I have to do something else?

Comment (FabioLuporini) [reply examples/seismic/tutorials/04_dask.ipynb]: Probably yes. I have asked reviews from 4 of us, but nobody replied yet.

Comment (navjotk) [reply examples/seismic/tutorials/04_dask.ipynb]: > for homogeneity with the other notebooks, I'd keep the original variable names

I agree with this

File: examples/seismic/tutorials/04_dask.ipynb [hunk 2]
```diff
@@ -324,11 +342,17 @@
...
-    "    # Set up solver.\n",
-    "    solver = AcousticWaveSolver(model0, geometry, space_order=4)\n",
+    "    # Set up solver -- load the solver used above in the genration of the syntethic data.    \n",
+    "    with open(\"arguments.pkl\", \"rb\") as cp_file:\n",
+    "        cp = pickle.load(cp_file)\n",
+    "        \n",
+    "    op = cp['operator']\n",
+    "    op.model=model0\n",
+    "    op.geometry=geometry\n",
+    "    op.dt=dt\n",
```
Comment (navjotk) [inline examples/seismic/tutorials/04_dask.ipynb]: I would like it better if the part about synthetic data was inside a self-contained `load_shot(i)` function/method that  can do synthetic data in the example, and can be replaced with actual loading from a file for a more realistic setup. What do other people think?

Reusing the solver that was used to generate the synthetic data, for the actual inversion, seems a bit backwards in that sense.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 3]
```diff
@@ -324,11 +342,17 @@
...
     "\n",
     "    # Compute simulated data and full forward wavefield u0\n",
-    "    d, u0, _ = solver.forward(save=True)\n",
+    "    d, u0, _ = op.forward(save=True)\n",
```
Comment (navjotk) [inline examples/seismic/tutorials/04_dask.ipynb]: Did we just name a `AcousticWaveSolver` object as `op`? That sounds confusing.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 4]
```diff
@@ -341,7 +365,7 @@
     "    # Compute gradient using the adjoint-state method. Note, this\n",
     "    # backpropagates the data misfit through the model.\n",
     "    grad = Function(name=\"grad\", grid=model0.grid)\n",
-    "    solver.gradient(rec=residual, u=u0, grad=grad)\n",
+    "    op.gradient(rec=residual, u=u0, grad=grad)\n",
```
Comment (navjotk) [inline examples/seismic/tutorials/04_dask.ipynb]: Same confusion. The solver object should be called solver. When I see op, I think Operator.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 5]
```diff
@@ -443,17 +467,61 @@
...
+      "Operator `initdamp` run in 0.01 s\n",
+      "Operator `padfunc` run in 0.01 s\n"
+     ]
+    },
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "      fun: 211.63605994835484\n",
+      "      fun: 203.91187255306352\n",
```
Comment (navjotk) [inline examples/seismic/tutorials/04_dask.ipynb]: Is the output expected to change?

File: examples/seismic/tutorials/04_dask.ipynb [hunk 6]
```diff
@@ -178,41 +180,40 @@
...
     "\n",
+    "def generate_shotdata(src_coor,rec_coor):\n",
+    "    true_model = get_true_model()\n",
     "    # Geometry \n",
-    "    geometry = AcquisitionGeometry(true_model, rec_coordinates, src_coordinates,\n",
+    "    geometry = AcquisitionGeometry(true_model, rec_coor, src_coor,\n",
     "                                   param['t0'], param['tn'], src_type='Ricker',\n",
     "                                   f0=param['f0'])\n",
-    "    # Set up solver.\n",
     "    solver = AcousticWaveSolver(true_model, geometry, space_order=4)\n",
```
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: This the only place where the `solver` is created. This same `solver` is used in the `generated_shotdata_i` and `fwi_gradient_i` functions.  It could be passed as an argument to the functions but I did not want to modify much the structure of the notebook so I created a dictionary with the `geometry` and the `solver` and then I used the `pickle` module. An “unpickling” operation is performed in the functions in order to use the `solver`.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 3 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: I can keep the old name to avoid confusion

File: examples/seismic/tutorials/04_dask.ipynb (hunk 4 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: The same thing here.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 7]
```diff
@@ -178,41 +180,40 @@
...
     "    \"\"\"\n",
-    "    true_model = get_true_model()\n",
-    "    shot_id = param['shot_id']\n",
-    "    \n",
-    "    src_coordinates = np.empty((1, len(param['shape'])))\n",
-    "    src_coordinates[0, :] = [30, param['shot_id']*1000./(param['nshots']-1)]\n",
+    "    # Reconstruct objects\n",
+    "    with open(\"arguments.pkl\", \"rb\") as cp_file:\n",
+    "        cp = pickle.load(cp_file)\n",
+    "        \n",
```
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: See the answer to your other comment below. https://github.com/opesci/devito/pull/945/files/243249b2ba50444bac7370dcd00789a6245ef90c#r331614314

File: examples/seismic/tutorials/04_dask.ipynb (hunk 2 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: The notebook would be much more organized I think, but as @ggorman ever said, the notebooks are given for illustrative purposes only. Someone could adapt it for your own use case and can do so better than us.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 5 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: I guess it is the last value of the objective function, so in principle, it should be the same. I assumed that the result was correct looking at the figures. The final models in both versions of the notebooks are virtually the same and I see absolutely no reason to get different values. Maybe @mloubout can help us understand a little more.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 8]
```diff
@@ -223,27 +224,43 @@
...
+      "Operator `Forward` run in 0.21 s\n",
+      "Operator `Forward` run in 0.20 s\n"
      ]
     }
    ],
    "source": [
     "#NBVAL_IGNORE_OUTPUT\n",
-    "\n",
-    "# Start Dask cluster\n",
+    "print('start')\n",
```
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: I can remove that print.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 9]
```diff
@@ -223,27 +224,43 @@
...
     "cluster = LocalCluster(n_workers=2, death_timeout=600)\n",
-    "client = Client(cluster)\n",
-    "\n",
+    "c = Client(cluster)\n",
+    "print('OK')\n",
```
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: idem

File: examples/seismic/tutorials/04_dask.ipynb (hunk 5 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: I ran several times the old notebook and got the same value as the new 203.91187255306352. I don't know where came from the other value. I confirmed the results are the same for both cases (old and new notebooks) OK.

Comment (navjotk) [reply examples/seismic/tutorials/04_dask.ipynb]: Ok this might have changed because of some previous change and the notebook wasn't rerun

File: examples/seismic/tutorials/04_dask.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (navjotk) [reply examples/seismic/tutorials/04_dask.ipynb]: In that case, firstly can we have the comment back? No reason to remove an explanatory comment, I guess?

File: examples/seismic/tutorials/04_dask.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (navjotk) [reply examples/seismic/tutorials/04_dask.ipynb]: Can we do it without writing and reading a file?

[the remark this answers is not in the document]
Comment (mloubout) [reply examples/seismic/tutorials/04_dask.ipynb]: Geometry is an attribute of the solver so should only need to save solver and get it via `solver.geometry`

File: examples/seismic/tutorials/04_dask.ipynb [hunk 10]
```diff
@@ -178,41 +180,40 @@
...
+    "    shot_id=param['shot_id']\n",
     "    \n",
-    "    # Number of receiver locations per shot.\n",
-    "    nreceivers = 101\n",
-    "\n",
-    "    # Set up receiver data and geometry.\n",
-    "    rec_coordinates = np.empty((nreceivers, len(param['shape'])))\n",
-    "    rec_coordinates[:, 1] = np.linspace(0, true_model.domain_size[0], num=nreceivers)\n",
-    "    rec_coordinates[:, 0] = 980. # 20m from the right end\n",
+    "    geometry.src_positions[0,:]=[30, shot_id*1000./(param['nshots']-1)]\n",
```
Comment (mloubout) [inline examples/seismic/tutorials/04_dask.ipynb]: what is the point of saving the Geometry if modified right after reading it?

**Comment by FabioLuporini** [comment]
@mloubout :

> But overall I think this is a way to "hide" a memry issue that the clear_cache PR is trying to fix in a better way IMO

it's a different issue: https://github.com/opesci/devito/issues/845

**Comment by navjotk** [comment]
> Now it is a memized function so it won't re-generate and re-compile anything but because you are on a different node the actual c-file does not exist.

This assumes no shared filesystem. I see no obvious reason why we shouldn't do this when we have a shared file system. 

There are scenarios where we have no choice but to do this - e.g. intel compiler license only available on login nodes. 

> But overall I think this is a way to "hide" a memry issue that the clear_cache PR is trying to fix in a better way IMO

The solver class has its own issues that are not related to `clear_cache`. This is avoiding those issues by not creating more than one solver. 

The idea of reusing `Operator` objects has been a key assumption in Devito for a long time but the examples all create new `Operator` objects very actively, so I think this modification is a step in the right direction.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: > Can we do it without writing and reading a file?

Yes, but it would have to be created a new `solver` at least one more time, i.e. in the `fwi_gradient` function. The new `solver` would be passed to `fwi_gradient_i` function to compute the gradient for each shot.  As this function runs in parallel (via Dask) the `solver` would have to be sent to all workers through a `broadcast` or loaded from a file. I chose the last option. I don't know if this would work on a cluster as @mloubout pointed out.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 10 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: Sure, you are right, there is no need to save the `geometry`. In this example, the geometry is the same, the only thing that changes is the source position. In this case, I could change the `dump_shot_data` function to save only the source positions instead of `geometry` and just to assign the `src_positions` to the `solver.geometry` in the `fwi_gradient_i` function.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb]: Yes, the comment has to stay.

**Comment by FabioLuporini** [comment]
@navjotk @mloubout 

can you provide a summary? where do we stand now this PR ?

File: examples/seismic/tutorials/04_dask.ipynb [hunk 11]
```diff
@@ -322,18 +320,23 @@
...
     "\n",
-    "    geometry, rec = load_shot_data(param['shot_id'], dt)\n",
-    "    geometry.model = model0\n",
-    "    # Set up solver.\n",
-    "    solver = AcousticWaveSolver(model0, geometry, space_order=4)\n",
+    "    # Get src_position and data\n",
+    "    src_positions, rec = load_shot_data(param['shot_id'], dt)\n",
+    "\n",
+    "    # Set attributes to solver\n",
+    "    solver.model=model0\n",
```
Comment (FabioLuporini) [inline examples/seismic/tutorials/04_dask.ipynb]: changing state from the outside world is bad and could have unpredictable side effects

one should instead `rebuild`

Comment (rhodrin) [reply examples/seismic/tutorials/04_dask.ipynb]: `Model` has a `vp.setter` wrapper (that we talked about before) so we should to able to use that.

**Comment by rhodrin** [comment]
@navjotk @mloubout @FabioLuporini:
Aside from obvious tidying that can be done, ensuring that object properties aren't being overwritten etc. what are the main things you feel need to be dealt with before this PR is mergeable?

**Comment by philippwitte** [comment]
Hi guys, 
Rhodri asked me to help out with this tutorial. I read through this thread and had a look at the current version, but can you summarize what the current issues are that are still outstanding?

I think the way the functions are currently set up to compute the functional value and gradient in parallel generally make sense. Namely setting up the solver inside each respective (serial) function call. I'm personally not a big fan of passing only a dictionary called `params` to the `fwi_gradient_i` function. Speaking from my own experience, I think these dictionaries tend to end up as a graveyard for all kinds of parameters and it's not always clear what is overwritten inside the function or what information is redundant. I think it's cleaner to have an objective function that takes a set of well defined input arguments, namely something like `observed_data, rec_geometry, src_geometry, src_wavelet, model`, which would be all the necessary information to evaluate the fwi objective function. But I understand it's a matter of personal taste and may not be something we want to change at this point.

Anyway, please let me know how I can help out!

**Comment by rhodrin** [comment]
> Hi guys,
> Rhodri asked me to help out with this tutorial. I read through this thread and had a look at the current version, but can you summarize what the current issues are that are still outstanding?
> 
> I think the way the functions are currently set up to compute the functional value and gradient in parallel generally make sense. Namely setting up the solver inside each respective (serial) function call. I'm personally not a big fan of passing only a dictionary called `params` to the `fwi_gradient_i` function. Speaking from my own experience, I think these dictionaries tend to end up as a graveyard for all kinds of parameters and it's not always clear what is overwritten inside the function or what information is redundant. I think it's cleaner to have an objective function that takes a set of well defined input arguments, namely something like `observed_data, rec_geometry, src_geometry, src_wavelet, model`, which would be all the necessary information to evaluate the fwi objective function. But I understand it's a matter of personal taste and may not be something we want to change at this point.
> 
> Anyway, please let me know how I can help out!

Hi @philippwitte ,

Thanks for agreeing to help out! So first of all, if there are any improvements or restructuring you'd like to add to the notebook please feel free to be creative and incorporate any adaptions you'd like to see.

Aside from that, in the current notebook there are various grammatical mistakes that need fixing and the instances where object properties are being overwritten need to be fixed etc.

With regards to the discussion between @navjotk and @mloubout, I'll try and speak with Navjot today and get back to you with any fixes/adaptations he'd like to see.

And to give you some context, this notebook will be used as the basis for an FWI 'challenge' for workshop participants who don't wish to get involved in the GPU hackathon.

**Comment by mloubout** [comment]
What's the plan on this, and considreing the new version of the Dask tuto @philippwitte made for oghpc

**Comment by mloubout** [comment]
So what's the plan for this one?

**Comment by FabioLuporini** [comment]
I don't even remember where we stand with this one :grimacing: 

I think we should first of all rebase it on latest master as there are conflicts with the current Dask tutorial. 

@mloubout @rhodrin this is what @ofmla originally wrote in the description of this PR:

> Now the Operator is created only once and it is reused for the modeling of all the shots and the gradient computation. In this way, there should not be memory issues.

going back to the very beginning -- do we all agree that this is a step we want to make to improve things?

**Comment by mloubout** [comment]
>do we all agree that this is a step we want to make to improve things?

I think it's worth a notebook, but i'd say a separate one as both ways have their merits. So I would say close this PR and open a new one with a new notebook.

**Comment by ofmla** [comment]
> > do we all agree that this is a step we want to make to improve things?
> 
> I think it's worth a notebook, but i'd say a separate one as both ways have their merits. So I would say close this PR and open a new one with a new notebook.

So, what you got in mind, @mloubout @FabioLuporini ? maybe I can help

**Comment by mloubout** [comment]
Probably just rename the notebook (like `..._dask_pickle_op.ipynb` and rebase would be a good start.

**Comment by ofmla** [comment]
@mloubout @FabioLuporini I rebased as suggested and created a new notebook 12_pickle_dask_op.ipynb with a brief introduction. I kept much of the original notebook, but I can add more explanatory comments If you want. The notebook deliver broadly the same results as 04_dask.ipynb.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 12]
```diff
@@ -205,9 +204,21 @@
      "nbval-ignore-output"
     ]
    },
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/muriloboratto/anaconda3/envs/devito/lib/python3.8/site-packages/distributed/node.py:151: UserWarning: Port 8787 is already in use.\n",
```
Comment (FabioLuporini) [inline examples/seismic/tutorials/04_dask.ipynb:212]: this should be dropped

File: examples/seismic/tutorials/12_pickle_dask_op.ipynb [hunk 13]
```diff
@@ -0,0 +1,649 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/muriloboratto/anaconda3/envs/devito/lib/python3.8/site-packages/distributed/node.py:151: UserWarning: Port 8787 is already in use.\n",
```
Comment (FabioLuporini) [inline examples/seismic/tutorials/12_pickle_dask_op.ipynb:217]: this should be dropped. Could you look for all messages like this and purge the notebook ?

**Comment by mloubout** [comment]
I think once @FabioLuporini comments are answered, should be GTG. Can you open an issue about the `src/rec_position` shape problem and assign it to me.

File: examples/seismic/tutorials/04_dask.ipynb (hunk 12 above)
Comment (ofmla) [reply examples/seismic/tutorials/04_dask.ipynb:212]: done

File: examples/seismic/tutorials/12_pickle_dask_op.ipynb (hunk 13 above)
Comment (ofmla) [reply examples/seismic/tutorials/12_pickle_dask_op.ipynb]: done

**Comment by ofmla** [comment]
> Following the convention used throughout devito/examples, we should name the new notebook as "04_dask_pickling.ipynb". And since it's an extension of 04_dask.ipynb, we should rip off most of the (now redundant) explanations, and add a short section at the beginning explaining what the delta between 04_dask and 04_dask_pickling is.

I renamed the tutorial and add a short paragraph that explains the differences between both tutorial  04_dask and 04_dask_pickling. I hope you like it, anyway, if more changes are needed let me know

> Said that, I super like the idea of moving this to a new notebook. I think we're right on track towards merging

File: examples/seismic/tutorials/04_dask.ipynb (hunk 12 above)
Comment (mloubout) [reply examples/seismic/tutorials/04_dask.ipynb:212]: It's still here.

**Comment by FabioLuporini** [comment]
Merged.
